### PR TITLE
disabledPushdown -> pushdownLevel

### PR DESCRIPTION
**Pull request by cryogenian**
More pushdown control

### CODE REVIEW

File: datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala [hunk 1]
```diff
@@ -20,14 +20,58 @@ import slamdata.Predef._
...
 import argonaut._, Argonaut._
 
+import cats.kernel.Order
+
 import com.mongodb.ConnectionString
 
 import quasar.physical.mongo.MongoResource.{Database, Collection}
 
+trait PushdownLevel extends Product with Serializable
+final case object Disabled extends PushdownLevel
```
Comment (wemrysi) [inline datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala]: We've been defining case constructors in the parent type's companion object by convention.

File: datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala [hunk 2]
```diff
@@ -20,14 +20,58 @@ import slamdata.Predef._
...
+      case Light => jString("light")
+      case Full => jString("full")
+    }
+  }
+  implicit val decodePushdownLevel: DecodeJson[PushdownLevel] = new DecodeJson[PushdownLevel] {
+    def decode(j: HCursor): DecodeResult[PushdownLevel] = j.as[String] flatMap {
+      case "disabled" => DecodeResult.ok(Disabled)
+      case "light" => DecodeResult.ok(Light)
+      case "full" => DecodeResult.ok(Full)
+      case _ => DecodeResult.fail("This is not PushdownLevel", CursorHistory(List()))
```
Comment (wemrysi) [inline datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala]: I think you can use the history from `j` here, instead of an empty one.

**Comment by cryogenian** [comment]
Moved #104 into this pr to minimize merge problems.

+ Extracted `safeField` and `safeCartouches` 
+ Used history from outside json 
+ `PushdownLevel` extracted into separate file, moved case classes into `PusdownLevel` object 

[ch5592] [ch5588]

**Comment by djspiewak** [comment]
@cryogenian Quick change to this menagerie of things. Can you default pushdown to **Off** for now and feature flag the drop-down selector in the frontend? We've had a request to make the rollout a little more controlled.

File: datasource/src/main/scala/quasar/physical/mongo/MongoExpression.scala [hunk 3]
```diff
@@ -140,8 +155,9 @@ object MongoExpression {
   }
 
   object Projection {
-    implicit val eqProjection: Eq[Projection] = new Eq[Projection] {
-      def eqv(a: Projection, b: Projection): Boolean = a.steps.toList === b.steps.toList
+    implicit val orderProjection: Order[Projection] = new Order[Projection] {
+      def compare(a: Projection, b: Projection): SInt =
+        Order[List[ProjectionStep]].compare(a.steps.toList, b.steps.toList)
```
Comment (wemrysi) [inline datasource/src/main/scala/quasar/physical/mongo/MongoExpression.scala]: :golf: `Order.orderBy(_.steps.toList)`

File: datasource/src/main/scala/quasar/physical/mongo/interpreter/Cartesian.scala [hunk 4]
```diff
@@ -22,18 +22,19 @@ import cats.syntax.traverse._
...
 import shims._
 
+import E.ProjectionStep._
+
 object Cartesian {
   def apply(
       uniqueKey: String,
       version: Version,
-      cartouches: Map[CPathField, (CPathField, List[ScalarStage.Focused])],
+      cartouches: Map[Field, (Field, List[ScalarStage.Focused])],
```
Comment (wemrysi) [inline datasource/src/main/scala/quasar/physical/mongo/interpreter/Cartesian.scala]: :+1:

**Pull request merged**